### DISCUSSION TITLE
Debian - Warnings about not installed Plugins

### DISCUSSION THREAD

**Discussion by juergenc** [rpm-software-management/mock#1194] [Q&A]
### Short Description
 I installed mock from source on Debian 12 (Bookworm) OS. Since for that OS level a Debian distribution package for mock does not exist anymore and is just on the wish list, I installed it from GitHub source here. In earlier Debian OS releases a `mock` installation package was still available. See also [Debian bug #1025460](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1025460)

### Output of `rpm -q mock`
Of course, there is none.

### Steps to reproduce issue
1. `git clone https://github.com/rpm-software-management/mock.git`
2.  I used this commit:
 ```
commit 8ad1f9ecbf6bcf241fe9a36028dd9410f8c5011b (HEAD -> main, origin/main, origin/HEAD)
Author: Pavel Raiskup <praiskup@redhat.com>
Date:   Thu Aug 10 11:11:58 2023 +0200

    Post-release administrivia
    
    Per https://rpm-software-management.github.io/mock/Releasing-Mock
 ```
3.  I installed all additional Debian packages needed by mock, e.g. `python-is-python3 python3-dev libpython3-dev python3-rpm dnf usermode python3-six python3-all-dev python3-distro python3-pyroute2 python3-backoff python3-jinja2 python3-requests python3-rpm util-linux coreutils procps-ng shadow-utils systemd-container podman`
4. I copied all files from mock repository to the right place for execution
5. I did all the necessary configuration stuff e.g. created a a group `mock` and added the build user to that group...
6. I started mock and managed to make it work, it builds several of my projects impeccably
7.  But with every build I see two or three times after every start of a stage these warning messages:
```
[...]
Start(bootstrap): init plugins
WARNING: root_cache plugin is enabled in configuration but is not installed
WARNING: yum_cache plugin is enabled in configuration but is not installed
WARNING: ccache plugin is enabled in configuration but is not installed
WARNING: mount plugin is enabled in configuration but is not installed
WARNING: bind_mount plugin is enabled in configuration but is not installed
WARNING: selinux plugin is enabled in configuration but is not installed
WARNING: package_state plugin is enabled in configuration but is not installed
WARNING: hw_info plugin is enabled in configuration but is not installed
Finish(bootstrap): init plugins      
Start: init plugins
WARNING: root_cache plugin is enabled in configuration but is not installed
WARNING: yum_cache plugin is enabled in configuration but is not installed
WARNING: ccache plugin is enabled in configuration but is not installed
WARNING: mount plugin is enabled in configuration but is not installed
WARNING: bind_mount plugin is enabled in configuration but is not installed
WARNING: selinux plugin is enabled in configuration but is not installed
WARNING: package_state plugin is enabled in configuration but is not installed
WARNING: hw_info plugin is enabled in configuration but is not installed
Finish: init plugins
[...]
```

Do not forget to mention full commandline with the mock command you executed.
```
mock -D "dist .$EL" -r "$EPEL_CFG" --configdir=/etc/mock --buildsrpm --sources="$SRC_DIR" --spec "$SRC_DIR/redhat/$NAME.spec" --resultdir=$RESULT_DIR --no-cleanup-after
mock -D "dist .$EL" -r "$EPEL_CFG" --configdir=/etc/mock --no-clean --rebuild "$RESULT_DIR/$NAME-$REDHAT_VERSION-1.$EL.src.rpm" --resultdir=$RESULT_DIR
```

where `$EPEL_CFG` can be:
- `oraclelinux-9-x86_64` (self defined)
- `oraclelinux-8-x86_64` (self defined)
- `oraclelinux-7-x86_64` (self defined)
- ... or some other (epel-x-...)

and `$EL` '7' or '8' or '9'

### Any additional notes
The `mockbuild/plugins` directory with the Python modules exists:
```
ls -la /usr/local/lib/python3.11/dist-packages/mockbuild/plugins/
total 144
drwxr-xr-x 2 jk users  4096 21. Aug 15:44 .
drwxr-xr-x 4 jk users  4096 21. Aug 20:20 ..
-rw-r--r-- 1 jk users  1778 21. Aug 15:44 bind_mount.py
-rw-r--r-- 1 jk users  2426 21. Aug 15:44 ccache.py
-rw-r--r-- 1 jk users  2728 21. Aug 15:44 chroot_scan.py
-rw-r--r-- 1 jk users  1391 21. Aug 15:44 compress_logs.py
-rw-r--r-- 1 jk users  1833 21. Aug 15:44 hw_info.py
-rw-r--r-- 1 jk users 16034 21. Aug 15:44 lvm_root.py
-rw-r--r-- 1 jk users  2076 21. Aug 15:44 mount.py
-rw-r--r-- 1 jk users 34011 21. Aug 15:44 overlayfs.py
-rw-r--r-- 1 jk users  3355 21. Aug 15:44 package_state.py
-rw-r--r-- 1 jk users  5281 21. Aug 15:44 pm_request.py
-rw-r--r-- 1 jk users  1450 21. Aug 15:44 procenv.py
-rw-r--r-- 1 jk users 10233 21. Aug 15:44 root_cache.py
-rw-r--r-- 1 jk users  4697 21. Aug 15:44 rpkg_preprocessor.py
-rw-r--r-- 1 jk users  3543 21. Aug 15:44 selinux.py
-rw-r--r-- 1 jk users  1335 21. Aug 15:44 showrc.py
-rw-r--r-- 1 jk users  1428 21. Aug 15:44 sign.py
-rw-r--r-- 1 jk users  3766 21. Aug 15:44 tmpfs.py
-rw-r--r-- 1 jk users  5098 21. Aug 15:44 yum_cache.py
```

What did I wrong? Did I miss something or screwed something up?

I want to make use of `ccache` and it does not work. But `ccache` is installed and configured on my Debian 12 build system:
```
ccache -v --show-stats
Cache directory:    /var/cache/mock/ccache
Config file:        /var/cache/mock/ccache/ccache.conf
System config file: /etc/ccache.conf
Stats updated:      never
Local storage:
  Cache size (GB):  0.00 / 4.00 ( 0.00%)
  Files:               0
  Hits:                0
  Misses:              0
  Reads:               0
  Writes:              0
```

In `/etc/mock/site-defaults.cfg` I added the following lines for `ccache`:
```
config_opts['plugin_conf']['ccache_enable'] = True
config_opts['plugin_conf']['ccache_opts'] = {}
config_opts['plugin_conf']['ccache_opts']['max_cache_size'] = '4G'
config_opts['plugin_conf']['ccache_opts']['compress'] = None
config_opts['plugin_conf']['ccache_opts']['dir'] = "{{cache_topdir}}/{{root}}/ccache/u{{chrootuid}}/"
```

I would be glad in case you are willing to help me here even it is a Debian system.

Thanks!

**Comment by juergenc** [comment]
`/usr/local/lib/python3.11/dist-packages/mockbuild/plugin.py` it is here in line 49 and 52:
```
 31     @traceLog()
 32     def init_plugins(self, buildroot):
 33         if self.already_initialized:
 34             return
 35         self.already_initialized = True
 36         for key in list(self.plugin_conf.keys()):
 37             if key.endswith('_opts'):
 38                 self.plugin_conf[key]['basedir'] = buildroot.basedir
 39                 self.plugin_conf[key]['cache_topdir'] = buildroot.cache_topdir
 40                 self.plugin_conf[key]['cachedir'] = buildroot.cachedir
 41                 self.plugin_conf[key]['root'] = buildroot.shared_root_name
 42                 self.plugin_conf[key]['resultdir'] = buildroot.resultdir
 43 
 44         self.state.start("init plugins")
 45         # Import plugins  (simplified copy of what yum does). Can add yum
 46         # features later when we prove we need them.
 47         for plugin in self.plugins:
 48             if self.plugin_conf.get("{0}_enable".format(plugin)):
 49                 spec = importlib.machinery.PathFinder.find_spec(plugin, [self.plugin_dir])
 50                 if not spec:
 51                     buildroot.root_log.warning(
 52                         "{0} plugin is enabled in configuration but is not installed".format(plugin))
 53                     continue
 54                 module = importlib.util.module_from_spec(spec)
 55                 spec.loader.exec_module(module)
 56                 sys.modules[spec.name] = module
 57 
 58                 if not hasattr(module, 'requires_api_version'):
 59                     raise Error('Plugin "%s" doesn\'t specify required API version' % plugin)
 60                 requested_api_version = module.requires_api_version
 61                 if requested_api_version != current_api_version:
 62                     raise Error('Plugin version mismatch - requested = %s, current = %s'
 63                                 % (requested_api_version, current_api_version))
 64 
 65                 run_in_bootstrap = getattr(module, "run_in_bootstrap", True)
 66                 # simplified (buildroot.is_bootstrap and run_in_bootstrap) or not buildroot.is_bootstrap
 67                 if not buildroot.is_bootstrap or run_in_bootstrap:
 68                     module.init(self, self.plugin_conf["{0}_opts".format(plugin)], buildroot)
 69         self.state.finish("init plugins")
```

Variable `self.plugin_dir` is initialised with path (surprise) `/usr/libexec/mock/mockbuild/plugins/`.
I created it and put a symlink there  `plugins -> /usr/local/lib/python3.11/dist-packages/mockbuild/plugins`.

Now, when I start another build I can see output like:
```
...
INFO: calling preinit hooks
INFO: enabled root cache
...
INFO: enabled ccache
INFO: enabled HW Info plugin
```

But now build fails because not finding a ccache package:
```
...
Finish: chroot init
Start: build phase for ehbackup-bkclientd-3.2.16.2.g00e0ac86b-1.el7.src.rpm
Start: build setup for ehbackup-bkclientd-3.2.16.2.g00e0ac86b-1.el7.src.rpm
Building target platforms: x86_64
Building for target x86_64
Wrote: /builddir/build/SRPMS/ehbackup-bkclientd-3.2.16.2.g00e0ac86b-1.el7.src.rpm
Finish: build setup for ehbackup-bkclientd-3.2.16.2.g00e0ac86b-1.el7.src.rpm
Finish: build phase for ehbackup-bkclientd-3.2.16.2.g00e0ac86b-1.el7.src.rpm
ERROR: Exception(/var/lib/jenkins/workspace/backup/packages/redhat/ehbackup-bkclientd-3.2.16.2.g00e0ac86b-1.el7.src.rpm) Config(oraclelinux-7-x86_64) 0 minutes 12 seconds
INFO: Results and/or logs in: /var/lib/jenkins/workspace/backup/packages/redhat
INFO: Cleaning up build root ('cleanup_on_failure=True')
Start: clean chroot
Finish: clean chroot
ERROR: Error: No packages found for:
  ccache
```

This does not makes sense to me. Debian `ccache` package is installed outside buildroot and configured. I added it as `BuildRequires: ccache` in the rpm.spec for my project which sounds weird to me having it inside the buildroot but gave it a try. Of course, with the same result: "ERROR: Error: No packages found for:  ccache"

Now, what's wrong?

By the way, I checked it on an Oracle-Linux-8 native build system with mock 1.4.16. Same here, there does not exist a directory  `/usr/libexec/mock/mockbuild/plugins/` just `/usr/libexec/mock/` and the plugins are in the Python libdir `/usr/lib/python3.6/site-packages/mockbuild/plugins/`.

Even if we look into [mock.spec](https://github.com/rpm-software-management/mock/blob/main/mock/mock.spec) we cannot find a clue that the plugin modules needed to be copied to `/usr/libexec/mock/mockbuild/plugins/`.

**Comment by xsuchy** [comment]
1) None of the upstream developers use Debian (hmm maybe @Conan-Kudo ) so this is terra incognita for us.

2) You basically needs to do step by step what is in %build and %install section https://github.com/rpm-software-management/mock/blob/main/mock/mock.spec#L145

Ad the plugins - they should go to:
```
cp -a py/mockbuild %{python_sitelib}/
```

On my F38 machine the plugins are in `/usr/lib/python3.11/site-packages/mockbuild/plugins`

As this is not issue per se, I will convert it to discussion.

**Comment by xsuchy** [comment]
> This does not makes sense to me. Debian ccache package is installed outside buildroot and configured. I added it as BuildRequires: cache 

The ccache is has to be installed in buildroot. The error means that mock cannot do that.  Ccache on host is actually ignored.

**Comment by juergenc** [comment]
Alright, thanks, with the ccache. Mock cannot do that,for whatever reason. The log says no packages found. So, probably I have to add some more package repo in the cfg-file. That's a good hint! Thanks, I'll check it.

**Comment by juergenc** [comment]
By the way here track down with the plugins:

Regarding the Python modules for the plugin, I respected the `mock.spec` and I did:
`cp -a py/mockbuild %{python_sitelib}/` meaning, I copied everything to my `site-lib` which is in Debian meanwhile `dist-packages`:
`cp -a py/mockbuild  /usr/local/lib/python3.11/dist-packages/mockbuild/`

 I instrumented the `/usr/local/lib/python3.11/dist-packages/mockbuild/plugin.py` module to see where the module path should be:

```
 51                     buildroot.root_log.warning(
 52                         "{0} plugin is enabled in configuration but is not installed: '{1}'".format(plugin, self.plugin_dir))
```

And, as described, the path `/usr/libexec/mock/mockbuild/plugins/` is shown in the log for `{1}` and not, as expected in `/usr/local/lib/python3.11/dist-packages/mockbuild/plugins`.

Line 25, I did not specify `config['plugin_dir']` in any configuration file, neither in `/etc/mock/site-defaults.cfg` nor in any build config file:

```
15     @traceLog()
 16     def __init__(self, config, state):
 17         self.config = config
 18         self._hooks = {}
 19         self.state = state
 20
 21         self.already_initialized = False
 22
 23         self.plugins = config['plugins']
 24         self.plugin_conf = config['plugin_conf']
 25         self.plugin_dir = config['plugin_dir']
```

`config['plugin_dir']` is initialised in line 120 of `mock/py/mockbuild/config.py`:
```
...
23 from .constants import MOCKCONFDIR, PKGPYTHONDIR, VERSION
...
114     # (global) plugins and plugin configs.
115     # ordering constraings: tmpfs must be first.
116     #    root_cache next.
117     #    after that, any plugins that must create dirs (yum_cache)
118     #    any plugins without preinit hooks should be last.
119     config_opts['plugins'] = PLUGIN_LIST
120     config_opts['plugin_dir'] = os.path.join(PKGPYTHONDIR, "plugins")
```

The global variable `PKGPYTHONDIR` is imported from `mock/py/mockbuild/constants.py`
```
  8 # all of the variables below are substituted by the build system
  9 VERSION = "5.0.1~ui12"
 10 SYSCONFDIR = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])), "..", "etc")
 11 PYTHONDIR = os.path.dirname(os.path.realpath(sys.argv[0]))
 12 PKGPYTHONDIR = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])), "mockbuild")
 13 MOCKCONFDIR = os.path.join(SYSCONFDIR, "mock")
 14 # end build system subs
````

So, obviously, line 12 concatenates the current working directory `CWD` of mock during execution time must be `/usr/libexec/mock/` and that's why it concatenates `self.plugin_dir` to `/usr/libexec/mock/mockbuild/plugins/`.

**Comment by juergenc** [comment]
Coooool, works now! I added the developer directory (EPEL) of the distro where the ccache package is obtainable. And now I have a real build boost as expected right after the compiler cache warm-up :)

The reason why I expected ccache outside bildroot is because for Debian Pbuilder (counter part to Mock for Debian distro) ccache is installed outside the buildroot.  The same ccache instance is used for all build environments on a build machine.

Thanks for the hint and willing to help!
